### ISSUE TITLE
Async process inside success event

### ISSUE THREAD

**Issue by iamzoh** [CacheControl/json-rules-engine#235]
I'm having some trouble leveraging rule-chaining to add some runtime facts based on facts already in the Almanac like described here https://github.com/CacheControl/json-rules-engine/blob/master/docs/almanac.md#retrieve-fact-values-when-handling-events

I've modified https://github.com/CacheControl/json-rules-engine/blob/master/examples/07-rule-chaining.js to provide an primitive example of what I'm trying to do
```
/*
 * This is an advanced example demonstrating rules that passed based off the
 * results of other rules by adding runtime facts.  It also demonstrates
 * accessing the runtime facts after engine execution.
 *
 * Usage:
 *   node ./examples/07-rule-chaining.js
 *
 * For detailed output:
 *   DEBUG=json-rules-engine node ./examples/07-rule-chaining.js
 */

require('colors');
const { Engine } = require('../dist').Engine

/**
 * Setup a new engine
 */
const engine = new Engine();

/**
 * Rule for identifying people who may like screwdrivers
 */
const drinkRule = {
  conditions: {
    all: [
      {
        fact: 'drinksOrangeJuice',
        operator: 'equal',
        value: true,
      },
      {
        fact: 'enjoysVodka',
        operator: 'equal',
        value: true,
      },
    ],
  },
  event: { type: 'drinks-screwdrivers' },
  priority: 10, // IMPORTANT!  Set a higher priority for the drinkRule, so it runs first
};
engine.addRule(drinkRule);

/**
 * Register listeners with the engine for rule success and failure
 */
let facts;
engine
  .on('success', async (event, almanac) => {
    console.log(`${facts.accountId + ' DID '.green}meet conditions for the ${event.type.underline} rule.`);
    return new Promise((resolve) => {
      setTimeout(() => {
        console.log('promise resolved');
        resolve();
      }, 0);
    });
  })
  .on('failure', (event) => {
    console.log(`${facts.accountId} did ${'NOT'.red} meet conditions for the ${event.type.underline} rule.`);
  });

// define fact(s) known at runtime
facts = {
  accountId: 'washington',
  drinksOrangeJuice: true,
  enjoysVodka: true,
  isSociable: true,
};
engine
  .run(facts) // first run, using washington's facts
  .then(() => {
    console.log('engine run completed');
  })
  .catch();

/*
 * OUTPUT:
 *
 * washington DID meet conditions for the drinks-screwdrivers rule.
 * engine run completed
 * promise resolved
 */
```

I would expect the success event to complete before the engine run finishes like so:
```
/*
 * OUTPUT:
 *
 * washington DID meet conditions for the drinks-screwdrivers rule.
 * promise resolved
 * engine run completed
 */
```

Here's a more concrete example based on my use case. I'm reading some facts from the Almanac, and I plan to assign a runtime fact based on those values.

```
const { Engine } = require('../dist').Engine

/**
 * Setup a new engine
 */
require('colors');

const engine = new Engine();

/**
 * Rule for identifying people who may like screwdrivers
 */
const drinkRule = {
  conditions: {
    all: [
      {
        fact: 'drinksOrangeJuice',
        operator: 'equal',
        value: true,
      },
      {
        fact: 'enjoysVodka',
        operator: 'equal',
        value: true,
      },
    ],
  },
  event: { type: 'drinks-screwdrivers' },
  priority: 10, // IMPORTANT!  Set a higher priority for the drinkRule, so it runs first
};
engine.addRule(drinkRule);

/**
 * Register listeners with the engine for rule success and failure
 */
let facts;
engine
  .on('success', async (event, almanac) => {
    almanac.addRuntimeFact('screwdriverAficionado', true);
    console.log(`${facts.accountId + ' DID '.green}meet conditions for the ${event.type.underline} rule.`);
    let someInput;
    someInput = await almanac.factValue('someInput');
    someInput = await almanac.factValue('someInput');
    someInput = await almanac.factValue('someInput');
    someInput = await almanac.factValue('someInput');
    someInput = await almanac.factValue('someInput');
    someInput = await almanac.factValue('someInput');
    someInput = await almanac.factValue('someInput');
    someInput = await almanac.factValue('someInput');
    someInput = await almanac.factValue('someInput');
    almanac.addRuntimeFact(`newFact`, someInput);
    console.log('resolved input', someInput);
  })
  .on('failure', (event) => {
    console.log(`${facts.accountId} did ${'NOT'.red} meet conditions for the ${event.type.underline} rule.`);
  });

// define fact(s) known at runtime
facts = {
  accountId: 'washington',
  drinksOrangeJuice: true,
  enjoysVodka: true,
  isSociable: true,
  someInput: [1, 2, 3],
};
engine
  .run(facts)
  .then((results) => {
    console.log('engine run completed');
    // console.log(results);
  })
  .catch();

/*
 * OUTPUT:
 *
 * washington DID meet conditions for the drinks-screwdrivers rule.
 * engine run completed
 * resolved input [ 1, 2, 3 ]
 */
```

And if I remove some of those async processes, then I get my expected behavior
```
...
engine
  .on('success', async (event, almanac) => {
    almanac.addRuntimeFact('screwdriverAficionado', true);
    console.log(`${facts.accountId + ' DID '.green}meet conditions for the ${event.type.underline} rule.`);
    let someInput;
    someInput = await almanac.factValue('someInput');
    almanac.addRuntimeFact(`newFact`, someInput);
    console.log('resolved input', someInput);
  })
...

/*
 * OUTPUT:
 *
 * washington DID meet conditions for the drinks-screwdrivers rule.
 * resolved input [ 1, 2, 3 ]
 * engine run completed
 */
```

**Comment by naveen-shetty** [comment]
Hi @CacheControl, @iamzoh & I are from the same team.

so just to add more details,  effectively this is how the current setup looks like 


```js

/******JSRE Engine Code ******/
const EventEmitter = require('events')
const myEmitter = new EventEmitter()

const run = () => {
   myEmitter.emit('success');
   console.log("done processing"); 
}




/******Engine Client side code ******/

const getDogs = () => {
  return new Promise(resolve => {
    setTimeout(() => {
      resolve(['Woof', 'Woof', 'Woof'])
    }, 0)
  })
}
myEmitter.on('success', async () => {
  const dogs = await getDogs()
  console.log(dogs);
})

run();
```

And we were (wrongly) expecting the output to be 

```
[ 'Woof', 'Woof', 'Woof' ]
done processing
```

whereas the actual output is 
```
done processing
[ 'Woof', 'Woof', 'Woof' ]
```

Given the behavior of NodeJs event emitter, I understand why it is behaving this way.  

The question is would you want to support async function in the event emitter listener. 

For our use, we wanted to read the fact value from almanac but we cannot because `almanac.factValue` is promisified.

**Comment by MattCopenhaver** [comment]
I've also experienced this issue.  I do think a fix would need a major version bump, so wondering what @CacheControl's opinion is here.

**Comment by CacheControl** [comment]
I definitely agree that the asynchronous nature of the almanac methods really make them impossible to be utilized effectively within the "classic" event emissions interfaces.

I'd support the breaking change to have the event emission to wait on any promise returned.

**Comment by CacheControl** [comment]
I've taken a crack at this in #237 - relatively easy fix since EventEmitter2 supports this out of the box. Comments welcome. I still have some docs to write and examples to tweak, but I think this is functionally code complete. Since this is a breaking change, I've got at least 1 other breaking change I'd like to into into this next release, so I'm shooting for a release in the next two weeks.

@iamzoh @naveen-shetty if you'd like to confirm the fix, you can `npm install json-rules-engine@6.0.0-alpha-2`.

**Comment by iamzoh** [comment]
Hey @CacheControl, I can confirm the fix works for my use case. Thank you.

I'll take a look at #237 to review as well.

**Comment by CacheControl** [comment]
This feature is now available in **v6.0.0**. See [changelog](https://github.com/CacheControl/json-rules-engine/blob/master/CHANGELOG.md#60--2020-12-22)
